pkg/rpc/nodedialer/nodedialer.go

```diff
@@ -220,71 +196,37 @@ func (n *Dialer) DialInternalClient(
 // checked.
-func (n *Dialer) dial(
+func dial[Conn io.Closer](
 	ctx context.Context,
+	resolver AddressResolver,
+	dialFn func(addr string, nodeID roachpb.NodeID, locality roachpb.Locality, class rpcbase.ConnectionClass) *rpc.Connection[Conn],
 	nodeID roachpb.NodeID,
-	addr net.Addr,
-	locality roachpb.Locality,
-	checkBreaker bool,
 	class rpcbase.ConnectionClass,
-) (*grpc.ClientConn, *rpc.BatchStreamPool, drpc.Conn, *rpc.DRPCBatchStreamPool, error) {
+	checkBreaker bool,
+) (Conn, *rpc.Connection[Conn], error) {
+	var nilConn Conn
 	const ctxWrapMsg = "dial"
-	// Don't trip the breaker if we're already canceled.
-	if ctxErr := ctx.Err(); ctxErr != nil {
-		return nil, nil, nil, nil, errors.Wrap(ctxErr, ctxWrapMsg)
-	}
-	rpcConn := n.rpcContext.GRPCDialNode(addr.String(), nodeID, locality, class)
-	connect := rpcConn.ConnectEx
-	if !checkBreaker {
-		connect = rpcConn.ConnectNoBreaker
-	}
-	conn, dconn, err := connect(ctx)
+
+	addr, locality, err := resolver(nodeID)
 	if err != nil {
-		// If we were canceled during the dial, don't trip the breaker.
-		if ctxErr := ctx.Err(); ctxErr != nil {
-			return nil, nil, nil, nil, errors.Wrap(ctxErr, ctxWrapMsg)
-		}
-		err = errors.Wrapf(err, "failed to connect to n%d at %v", nodeID, addr)
-		return nil, nil, nil, nil, err
+		return nilConn, nil, err
 	}
-	pool := rpcConn.BatchStreamPool()
-	drpcStreamPool := rpcConn.DRPCBatchStreamPool()
-	return conn, pool, dconn, drpcStreamPool, nil
-}
-
-// drcpDial performs the dialing of the remote connection. If checkBreaker
-// is set (which it usually is), circuit breakers for the peer will be
-// checked. This method is similar to dial, but it dials a DRPC
-// connection instead of a gRPC connection.
-func (n *Dialer) drpcDial(
-	ctx context.Context,
-	nodeID roachpb.NodeID,
-	addr net.Addr,
-	locality roachpb.Locality,
-	checkBreaker bool,
-	class rpcbase.ConnectionClass,
-) (drpc.Conn, *rpc.DRPCBatchStreamPool, error) {
-	const ctxWrapMsg = "drpcDial"
 	// Don't trip the breaker if we're already canceled.
 	if ctxErr := ctx.Err(); ctxErr != nil {
-		return nil, nil, errors.Wrap(ctxErr, ctxWrapMsg)
+		return nilConn, nil, errors.Wrap(ctxErr, ctxWrapMsg)
 	}
-	rpcConn := n.rpcContext.DRPCDialNode(addr.String(), nodeID, locality, class)
-	connect := rpcConn.ConnectEx
+
+	log.VEventf(ctx, 2, "sending request to %s", addr)
+	rpcConn := dialFn(addr.String(), nodeID, locality, class)
+	connect := rpcConn.Connect
 	if !checkBreaker {
 		connect = rpcConn.ConnectNoBreaker
 	}
-	_, dconn, err := connect(ctx)
+	conn, err := connect(ctx)
 	if err != nil {
-		// If we were canceled during the dial, don't trip the breaker.
-		if ctxErr := ctx.Err(); ctxErr != nil {
-			return nil, nil, errors.Wrap(ctxErr, ctxWrapMsg)
-		}
-		err = errors.Wrapf(err, "failed to connect to n%d at %v", nodeID, addr)
-		return nil, nil, err
+		return nilConn, nil, errors.Wrapf(err, "failed to connect to n%d at %v", nodeID, addr)
```

Comment:
Not necessarily for this PR, but it would be good if we could tell whether drpc or gRPC were used in the error message.